File: algo/147. Insertion Sort List.js
// Linked List

/**
 * Definition for singly-linked list.
 * function ListNode(val) {
 *     this.val = val;
 *     this.next = null;
 * }
 */
/**
 * @param {ListNode} head
 * @return {ListNode}
 */
var insertionSortList = function(head) {
    var insert = function(toInsert) {
        var p1 = sortedStart;
        var p2 = sortedStart.next;
        while (p2 !== null && p2.val < toInsert.val) {
            p1 = p1.next;
            p2 = p2.next;
        }
        if (p1 === sortedEnd) {
            p1.next = toInsert;
            sortedEnd = toInsert;
        } else {
            p1.next = toInsert;
            toInsert.next = p2;
        }
    };
    if (head === null) {
        return head;
    }
    var dummy = new ListNode(-Infinity);
    dummy.next = head;
    var cursor = head;
    var sortedStart = dummy;
    var sortedEnd = dummy;
    
    while (cursor !== null) {
        var temp = cursor.next;
        cursor.next = null;
        insert(cursor, sortedStart, sortedEnd);
        cursor = temp;
    }
    return dummy.next;
};

File: round2/147. Insertion Sort List.cs
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     public int val;
 *     public ListNode next;
 *     public ListNode(int x) { val = x; }
 * }
 */
public class Solution {
    public ListNode InsertionSortList(ListNode head) {
        if (head == null) return null;
        ListNode dummy = new ListNode(0);
        dummy.next = head;
        ListNode cursor1, cursor2 = head.next;
        head.next = null;
        while (cursor2 != null) { 
            cursor1 = dummy;
            ListNode tmp = cursor2.next;
            while (cursor1 != null && cursor1.next != null && cursor2.val >= cursor1.next.val) {
                //Console.WriteLine(cursor1.next.val);
                cursor1 = cursor1.next;
            }
            cursor2.next = cursor1.next;
            cursor1.next = cursor2;
            cursor2 = tmp;
        }
        return dummy.next;
    }
}